bson/bsoncodec/time_codec.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) MongoDB, Inc. 2017-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bsoncodec
+
+import (
+	"fmt"
+	"reflect"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson/bsonoptions"
+	"go.mongodb.org/mongo-driver/bson/bsonrw"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+)
+
+const (
+	timeFormatString = "2006-01-02T15:04:05.999Z07:00"
+)
+
+var defaultTimeCodec = &TimeCodec{}
+
+// TimeCodec is the Codec used for struct values.
+type TimeCodec struct {
+	UseLocalTimeZone bool
+	DecodeFromString bool
+}
+
+var _ ValueCodec = &TimeCodec{}
+
+// NewTimeCodec returns a TimeCodec with options opts.
+func NewTimeCodec(opts ...*bsonoptions.TimeCodecOptions) (*TimeCodec, error) {
+	timeOpt := bsonoptions.MergeTimeCodecOptions(opts...)
+
+	codec := TimeCodec{}
+	if timeOpt.UseLocalTimeZone != nil {
+		codec.UseLocalTimeZone = *timeOpt.UseLocalTimeZone
+	}
+	if timeOpt.DecodeFromString != nil {
+		codec.DecodeFromString = *timeOpt.DecodeFromString
+	}
+	return &codec, nil
+}
+
+// DecodeValue is the ValueDecoderFunc for time.Time.
+func (tc *TimeCodec) DecodeValue(dc DecodeContext, vr bsonrw.ValueReader, val reflect.Value) error {
+	if !val.CanSet() || val.Type() != tTime {
+		return ValueDecoderError{Name: "TimeDecodeValue", Types: []reflect.Type{tTime}, Received: val}
+	}
+
+	var timeVal time.Time
+	valueType := vr.Type()
+	switch valueType {
+	case bsontype.DateTime:
+		// default logic
+		dt, err := vr.ReadDateTime()
+		if err != nil {
+			return err
+		}
+		timeVal = time.Unix(dt/1000, dt%1000*1000000)
+	case bsontype.String:
```

Comment:
Something I didn't catch on my first pass: maybe decoding a BSON string into a `time.Time` can always be OK? I don't see much value in returning an error anyway. The general robustness principle is relaxed with input and strict with output. We will never marshal a `time.Time` into a string but I don't see any harm in unmarshalling. I think this will make a better user experience overall and reduce the number of extra options we have to add and document in this epic.

Reply:
This sounds a lot like a we need a more general discussion about what parts of mgo's more lax decoding behavior, if any, we want to include in our default.

Reply:
My rule of thumb is "if we error while trying to unmarshal BSON type A into Go type B but mgo allows this", we should add this behavior without adding an option.

Reply:
done